pyro/distributions/naf.py

```diff
@@ -0,0 +1,115 @@
+    >>> naf = DeepSigmoidalFlow(arn, hidden_units=16)
+    >>> naf_module = pyro.module("my_naf", naf)
+    >>> naf_dist = dist.TransformedDistribution(base_dist, [naf])
+    >>> naf_dist.sample()  # doctest: +SKIP
+        tensor([-0.4071, -0.5030,  0.7924, -0.2366, -0.2387, -0.1417,  0.0868,
+                0.1389, -0.4629,  0.0986])
+
+    The inverse operation is not implemented. This would require numerical inversion, e.g., using a
+    root finding method - a possibility for a future implementation.
+
+    :param autoregressive_nn: an autoregressive neural network whose forward call returns a tuple of three
+        real-valued tensors, whose last dimension is the input dimension, and whose penultimate dimension
+        is equal to hidden_units.
+    :type autoregressive_nn: nn.Module
+    :param hidden_units: the number of hidden units to use in the NAF transformation (see Eq (8) in reference)
+    :type hidden_units: int
+
+    Reference:
+
+    Neural Autoregressive Flows [arXiv:1804.00779]
+    Chin-Wei Huang, David Krueger, Alexandre Lacoste, Aaron Courville
+
+    """
+
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+    event_dim = 1
+
+    def __init__(self, autoregressive_nn, hidden_units=16):
+        super(DeepSigmoidalFlow, self).__init__(cache_size=1)
+        self.arn = autoregressive_nn
+        self.hidden_units = hidden_units
+
+        # Not entirely sure this is necessary, but copying from NAF paper's implementation
+        self.safe_log = lambda x: torch.log(x * 1e2) - math.log(1e2)
+        self.safe_logit = lambda x: self.safe_log(x) - self.safe_log(1 - x)
+
+        # The output of safe_sigmoid is between [0.5*eps, 1-0.5*eps], which stops logit from returning +-Inf (overflow)
+        self.sigmoid = nn.Sigmoid()
+        self.safe_sigmoid = lambda x: self.sigmoid(x) * (1 - eps) + 0.5 * eps
+        self.logsoftmax = nn.LogSoftmax(dim=-2)
+
+    def _call(self, x):
+        """
+        :param x: the input into the bijection
+        :type x: torch.Tensor
+
+        Invokes the bijection x=>y; in the prototypical context of a TransformedDistribution `x` is a
+        sample from the base distribution (or the output of a previous flow)
+        """
+        # A, W, b ~ batch_shape x hidden_units x event_shape
+        A, W, b = self.arn(x)
+
+        # Divide the autoregressive output into the component activations
+        A = F.softplus(A)
+        W = F.softmax(W, dim=-2)
+
+        # The use of a special sigmoid here is so that logit doesn't overflow
+        # NOTE: Element-wise multiplication by W then summing over second-last dim is equivalent to
+        # dot-product over cols (or rows?) and cols (or rows?) of sigmoid term
+        # The unsqueeze on `x` is so that A * x.unsqueeze(-2) broadcasts correctly over hidden_units dimension
+        y = self.safe_logit(torch.sum(W * self.safe_sigmoid(A * x.unsqueeze(-2) + b), dim=-2))
```

Comment:
actually you'll want  `'...abc,...abc->...ac'`

Reply:
It's actually `'...bc,...bc->...c'` since it doesn't have to have a batch dimension. I did a test on the forward operation and it was up to twice as fast using the naive non-`torch.einsum` method for the forward operation (I presume autograd is the same speed for both). This is true even increasing the dimension of the NAF hidden units

Reply:
ok. i guess einsum still needs more perf enhancements....